lib/utils/mcputils/protocol.go

```diff
@@ -0,0 +1,174 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package mcputils
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/gravitational/trace"
+	"github.com/mark3labs/mcp-go/mcp"
+
+	apievents "github.com/gravitational/teleport/api/types/events"
+)
+
+// Type definitions from both mcp-go/client/transport or mcp-go are not suitable
+// for our reverse proxy use, thus this file redefines them.
+//
+// TODO(greedy52) switch to official golang lib or official go SDK if they offer
+// the level of handling we need. Same goes for other helpers like StdioXXX.
+
+// IDToString converts ID to its string representation.
+// ID should only be string, number, or nil.
+func IDToString(id mcp.RequestId) string {
+	if idStr, ok := id.(string); ok {
+		return idStr
+	}
+	return fmt.Sprintf("%v", id)
+}
+
+// JSONRPCParams defines params for request or notification.
+// TODO(greedy52) handle metadata
+type JSONRPCParams map[string]any
+
+// GetEventParams returns the apievents.Struct for auditing.
+func (p JSONRPCParams) GetEventParams() *apievents.Struct {
+	if p == nil {
+		return nil
+	}
+
+	eventParams, _ := apievents.EncodeMap(p)
+	return eventParams
+}
+
+// GetName returns the "name" param.
+func (p JSONRPCParams) GetName() (string, bool) {
+	if p == nil {
+		return "", false
+	}
+	name, ok := p["name"].(string)
+	return name, ok
+}
+
+// baseJSONRPCMessage is a base message that includes all fields for MCP
+// protocol.
+//
+// Note that json.RawMessage is used to keep the original content when
```

Comment:
nit: Another way to describe this is that for any user-defined field (and without affecting protocol behaviors), we keep it as `json.RawMessage` so that the function that knows its format can unmarshal it.